qiskit_experiments/data_processing/base.py

```diff
@@ -0,0 +1,99 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Defines the steps that can be used to analyse data."""
+
+from abc import ABCMeta, abstractmethod
+from typing import Any, Dict, List
+
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataAction(metaclass=ABCMeta):
+    """
+    Abstract action which is a single action done on measured data to process it.
+    Each subclass of DataAction must define the type of data that it accepts as input
+    using decorators.
+    """
+
+    # Key under which the node will output the data.
+    __node_output__ = None
+
+    def __init__(self):
+        """Create new data analysis routine."""
+        self._child = None
+        self._accepted_inputs = []
+
+    @property
+    def node_inputs(self) -> List[str]:
+        """Returns a list of input data that the node can process."""
+        return self._accepted_inputs
+
+    def add_accepted_input(self, data_key: str):
+        """
+        Allows users to add an accepted input data format to this DataAction.
+
+        Args:
+            data_key: The key that the data action will require in the input data dict.
+        """
+        self._accepted_inputs.append(data_key)
+
+    @abstractmethod
+    def process(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Applies the data processing step to the data.
+
+        Args:
+            data: the data to which the data processing step will be applied.
+
+        Returns:
+            processed data: The data that has been processed.
+        """
+
+    def check_required(self, data: Dict[str, Any]):
+        """Checks that the given data contains the right key.
+
+        Args:
+            data: The data to check for the correct keys.
+
+        Raises:
+            DataProcessorError: if the key is not found.
+        """
+        for key in data.keys():
+            if key in self.node_inputs:
+                return
+
+        raise DataProcessorError(f"None of {self.node_inputs} are in the given data.")
+
+    def format_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
@chriseclectic Can you please give a few examples of how this will look like from a user point of view?

Reply:
That's a good point. It will allow the framework to be more generic. With this we will be able to do
```python
data_processor = DataProcessor(...)
new_data = data_processor(data)
```
instead of 
```python
data_processor = DataProcessor(...)
new_data = data_processor.format(data)
```
This means that an analysis function could get as argument a `Callable` which could be a data processor but does not have to be.

Reply:
See here ca2d365

Reply:
Would it also work if the analysis function received `data_processor.format` as an argument?